An exceptionCaught() event was fired, and it reached at the tail of the pipeline.

Hello, our team are ready to use token-based authentication in sending APNs push notifications. But we got some problems when we use pushy in our test environment. Here are some debug logs:
```
2017-06-30 00:56:25,899 [nioEventLoopGroup-2-2] DEBUG [c.turo.pushy.apns.ApnsClientHandler] -
        Closing channel due to ping timeout.
----
2017-06-30 00:56:25,910 [nioEventLoopGroup-2-2] DEBUG [com.turo.pushy.apns.ApnsClient] -
        Disconnected. Next automatic reconnection attempt in 1 seconds.
----
2017-06-30 00:56:26,912 [nioEventLoopGroup-2-2] DEBUG [com.turo.pushy.apns.ApnsClient] -
        Attempting to reconnect.
----
2017-06-30 00:56:28,364 [nioEventLoopGroup-2-3] INFO  [com.turo.pushy.apns.ApnsClient] -
        Connection to api.development.push.apple.com/17.188.138.73:443 restored.
----
2017-06-30 00:57:47,766 [nioEventLoopGroup-2-1] WARN  [i.n.channel.DefaultChannelPipeline] -
        An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
----
java.io.IOException: Connection timed out
        at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
        at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
        at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
        at sun.nio.ch.IOUtil.read(IOUtil.java:192)
        at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:379)
        at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:288)
        at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1100)
        at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:372)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:123)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        at java.lang.Thread.run(Thread.java:745)
2017-06-30 00:57:47,767 [nioEventLoopGroup-2-1] DEBUG [com.turo.pushy.apns.ApnsClient] -
        Disconnected. Next automatic reconnection attempt in 1 seconds.
2017-06-30 00:57:48,769 [nioEventLoopGroup-2-1] DEBUG [com.turo.pushy.apns.ApnsClient] -
        Attempting to reconnect.
----
2017-06-30 00:57:50,215 [nioEventLoopGroup-2-4] INFO  [com.turo.pushy.apns.ApnsClient] -
        Connection to api.development.push.apple.com/17.188.138.73:443 restored.
----
2017-06-30 00:57:52,376 [nioEventLoopGroup-2-1] DEBUG [c.turo.pushy.apns.ApnsClientHandler] -
        Closing channel due to ping timeout.
----
2017-06-30 01:05:09,569 [nioEventLoopGroup-2-4] WARN  [i.n.channel.DefaultChannelPipeline] -
        An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
----
java.io.IOException: Connection timed out
        at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
        at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
        at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
        at sun.nio.ch.IOUtil.read(IOUtil.java:192)
        at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:379)
        at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:288)
        at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1100)
        at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:372)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:123)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        at java.lang.Thread.run(Thread.java:745)
2017-06-30 01:05:09,571 [nioEventLoopGroup-2-4] DEBUG [com.turo.pushy.apns.ApnsClient] -
        Disconnected. Next automatic reconnection attempt in 1 seconds.
----
2017-06-30 01:05:10,572 [nioEventLoopGroup-2-4] DEBUG [com.turo.pushy.apns.ApnsClient] -
        Attempting to reconnect.
----
2017-06-30 01:05:15,312 [nioEventLoopGroup-2-1] INFO  [com.turo.pushy.apns.ApnsClient] -
        Connection to api.development.push.apple.com/17.188.138.73:443 restored.
----
```
It seems that the reconnection is not stable when ApnsClient touches off the reconnection. The exception always happens after the first reconnection which happens in mid night when no one use the pushy server for some hours. And after first exception happens, the same exception will present occasionally.At the same time, other exception also can be present, like:
```
2017-06-30 03:19:57,036 [nioEventLoopGroup-2-2] DEBUG [c.turo.pushy.apns.ApnsClientHandler] -
        Closing channel due to ping timeout.
----
2017-06-30 03:19:57,039 [nioEventLoopGroup-2-2] DEBUG [com.turo.pushy.apns.ApnsClient] -
        Disconnected. Next automatic reconnection attempt in 1 seconds.
----
2017-06-30 03:19:58,041 [nioEventLoopGroup-2-2] DEBUG [com.turo.pushy.apns.ApnsClient] -
        Attempting to reconnect.
----
2017-06-30 03:20:09,304 [nioEventLoopGroup-2-3] WARN  [i.n.h.s.ApplicationProtocolNegotiationHandler] -
        [id: 0x07034b62, L:/121.43.60.132:13110 - R:api.development.push.apple.com/17.188.138.73:443] TLS handshake failed:
----
javax.net.ssl.SSLException: handshake timed out
        at io.netty.handler.ssl.SslHandler.handshake(...)(Unknown Source)
2017-06-30 03:20:09,307 [nioEventLoopGroup-2-3] INFO  [com.turo.pushy.apns.ApnsClient] -
        Failed to connect.
----
java.lang.IllegalStateException: Channel closed before HTTP/2 preface completed.
        at com.turo.pushy.apns.ApnsClient$3.operationComplete(ApnsClient.java:395)
        at com.turo.pushy.apns.ApnsClient$3.operationComplete(ApnsClient.java:387)
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:507)
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:481)
        at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:420)
        at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:104)
        at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:82)
        at io.netty.channel.AbstractChannel$CloseFuture.setClosed(AbstractChannel.java:1058)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.doClose0(AbstractChannel.java:686)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.close(AbstractChannel.java:664)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.close(AbstractChannel.java:607)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.close(DefaultChannelPipeline.java:1276)
        at io.netty.channel.AbstractChannelHandlerContext.invokeClose(AbstractChannelHandlerContext.java:624)
        at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:608)
        at io.netty.handler.ssl.SslHandler$8.operationComplete(SslHandler.java:1657)
        at io.netty.handler.ssl.SslHandler$8.operationComplete(SslHandler.java:1646)
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:507)
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:481)
        at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:420)
        at io.netty.util.concurrent.DefaultPromise.addListener(DefaultPromise.java:163)
        at io.netty.channel.DefaultChannelPromise.addListener(DefaultChannelPromise.java:93)
        at io.netty.channel.DefaultChannelPromise.addListener(DefaultChannelPromise.java:28)
        at io.netty.handler.ssl.SslHandler.safeClose(SslHandler.java:1646)
        at io.netty.handler.ssl.SslHandler.closeOutboundAndChannel(SslHandler.java:1457)
        at io.netty.handler.ssl.SslHandler.close(SslHandler.java:613)
        at io.netty.channel.AbstractChannelHandlerContext.invokeClose(AbstractChannelHandlerContext.java:624)
        at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:608)
        at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:465)
        at io.netty.handler.ssl.ApplicationProtocolNegotiationHandler.handshakeFailure(ApplicationProtocolNegotiationHandler.java:116)
        at io.netty.handler.ssl.ApplicationProtocolNegotiationHandler.userEventTriggered(ApplicationProtocolNegotiationHandler.java:94)
        at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:329)
        at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:315)
        at io.netty.channel.AbstractChannelHandlerContext.fireUserEventTriggered(AbstractChannelHandlerContext.java:307)
        at io.netty.handler.ssl.SslUtils.notifyHandshakeFailure(SslUtils.java:246)
        at io.netty.handler.ssl.SslHandler.notifyHandshakeFailure(SslHandler.java:1415)
        at io.netty.handler.ssl.SslHandler.access$1200(SslHandler.java:160)
        at io.netty.handler.ssl.SslHandler$5.run(SslHandler.java:1584)
        at io.netty.util.concurrent.PromiseTask$RunnableAdapter.call(PromiseTask.java:38)
at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:120)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:462)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        at java.lang.Thread.run(Thread.java:745)
2017-06-30 03:20:09,307 [nioEventLoopGroup-2-3] DEBUG [com.turo.pushy.apns.ApnsClient] -
        Disconnected. Next automatic reconnection attempt in 1 seconds.
----
```
and
```
2017-06-30 03:32:06,705 [nioEventLoopGroup-2-3] DEBUG [com.turo.pushy.apns.ApnsClient] -
        Attempting to reconnect.
----
2017-06-30 03:32:13,709 [nioEventLoopGroup-2-4] INFO  [com.turo.pushy.apns.ApnsClient] -
        Failed to connect.
----
io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection timed out: api.development.push.apple.com/17.188.138.73:443
        at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
        at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:739)
        at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:352)
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:340)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:632)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:579)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:496)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.ConnectException: Connection timed out
        ... 11 common frames omitted
2017-06-30 03:32:13,710 [nioEventLoopGroup-2-4] DEBUG [com.turo.pushy.apns.ApnsClient] -
        Disconnected. Next automatic reconnection attempt in 1 seconds.
----
```
Our pushy server create two ApnsClient which are used to push notification in turn. I count the exception more than 10 times last night. Another interesting thing is all kinds of exceptions mostly happen in 0-8 o'clock(GMT+8) for a few days. Finally, we are in China. Wish to your reply.

--- a/pushy/src/main/java/com/turo/pushy/apns/ApnsChannelFactory.java
+++ b/pushy/src/main/java/com/turo/pushy/apns/ApnsChannelFactory.java
@@ -66,6 +66,17 @@ class ApnsChannelFactory implements PooledObjectFactory<Channel>, Closeable {
     private static final AttributeKey<Promise<Channel>> CHANNEL_READY_PROMISE_ATTRIBUTE_KEY =
             AttributeKey.valueOf(ApnsChannelFactory.class, "channelReadyPromise");
 
+    @ChannelHandler.Sharable
+    private static class ConnectionNegotiationErrorHandler extends ChannelHandlerAdapter {
+
+        static final ConnectionNegotiationErrorHandler INSTANCE = new ConnectionNegotiationErrorHandler();
+
+        @Override
+        public void exceptionCaught(final ChannelHandlerContext context, final Throwable cause) {
+            context.channel().attr(CHANNEL_READY_PROMISE_ATTRIBUTE_KEY).get().tryFailure(cause);
+        }
+    }
+
     ApnsChannelFactory(final SslContext sslContext, final ApnsSigningKey signingKey,
                        final ProxyHandlerFactory proxyHandlerFactory, final int connectTimeoutMillis,
                        final long idlePingIntervalMillis, final long gracefulShutdownTimeoutMillis,
@@ -133,6 +144,7 @@ class ApnsChannelFactory implements PooledObjectFactory<Channel>, Closeable {
 
                             pipeline.addLast(new IdleStateHandler(idlePingIntervalMillis, 0, 0, TimeUnit.MILLISECONDS));
                             pipeline.addLast(apnsClientHandler);
+                            pipeline.remove(ConnectionNegotiationErrorHandler.INSTANCE);
 
                             channel.attr(CHANNEL_READY_PROMISE_ATTRIBUTE_KEY).get().trySuccess(channel);
                         } else {
@@ -142,6 +154,7 @@ class ApnsChannelFactory implements PooledObjectFactory<Channel>, Closeable {
                 });
 
                 pipeline.addLast(sslHandler);
+                pipeline.addLast(ConnectionNegotiationErrorHandler.INSTANCE);
             }
         });
     }

--- a/pushy/src/main/java/com/turo/pushy/apns/server/BaseHttp2Server.java
+++ b/pushy/src/main/java/com/turo/pushy/apns/server/BaseHttp2Server.java
@@ -32,6 +32,8 @@ import io.netty.handler.ssl.SslContext;
 import io.netty.handler.ssl.SslHandler;
 import io.netty.util.ReferenceCounted;
 import io.netty.util.concurrent.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import javax.net.ssl.SSLSession;
 import java.util.concurrent.atomic.AtomicBoolean;
@@ -44,6 +46,19 @@ abstract class BaseHttp2Server {
     private final boolean shouldShutDownEventLoopGroup;
 
     private ChannelGroup allChannels;
+
+    private static final Logger log = LoggerFactory.getLogger(BaseHttp2Server.class);
+
+    @ChannelHandler.Sharable
+    private static class ConnectionNegotiationErrorHandler extends ChannelHandlerAdapter {
+
+        static final ConnectionNegotiationErrorHandler INSTANCE = new ConnectionNegotiationErrorHandler();
+
+        @Override
+        public void exceptionCaught(final ChannelHandlerContext context, final Throwable cause) {
+            log.debug("Server caught an exception before establishing an HTTP/2 connection.", cause);
+        }
+    }
 
     BaseHttp2Server(final SslContext sslContext, final EventLoopGroup eventLoopGroup) {
 
@@ -70,12 +85,19 @@ abstract class BaseHttp2Server {
             protected void initChannel(final SocketChannel channel) {
                 final SslHandler sslHandler = sslContext.newHandler(channel.alloc());
                 channel.pipeline().addLast(sslHandler);
+                channel.pipeline().addLast(ConnectionNegotiationErrorHandler.INSTANCE);
 
                 sslHandler.handshakeFuture().addListener(new GenericFutureListener<Future<Channel>>() {
                     @Override
                     public void operationComplete(final Future<Channel> handshakeFuture) throws Exception {
-                        BaseHttp2Server.this.addHandlersToPipeline(sslHandler.engine().getSession(), channel.pipeline());
-                        BaseHttp2Server.this.allChannels.add(channel);
+                        if (handshakeFuture.isSuccess()) {
+                            BaseHttp2Server.this.addHandlersToPipeline(sslHandler.engine().getSession(), channel.pipeline());
+                            channel.pipeline().remove(ConnectionNegotiationErrorHandler.INSTANCE);
+
+                            BaseHttp2Server.this.allChannels.add(channel);
+                        } else {
+                            log.debug("TLS handshake failed.", handshakeFuture.cause());
+                        }
                     }
                 });
             }

